### PR TITLE
Add `LogSegmentBuilder` to construct `LogSegment`s

### PR DESCRIPTION
**Pull request by OussamaSaoudi-db**
## What changes are proposed in this pull request?

This pull request builds on #438 by creating a `LogSegmentBuilder`. This moves all logic for building a `LogSegment` out of `Snapshot`. This change is made in anticipation of `TableChanges`, which will represent CDFs and must construct its own `LogSegment`. 

The builder allows you to specify the following:
- A starting checkpoint
- The order of the commit files
- The start version 
- The end version
- Whether checkpoint files are omitted from the `LogSegment`.

## How was this change tested?

New tests are added to check the following:
1. Speciying start_version and start_checkpoint results in an error
2. Non contiguous commit files should result in an error
3. commit file ordering is correct in ascending and descending cases
4. Checkpoint files are correctly omitted if `with_omit_checkpoint_files` is specified.
5. start and end version work.

### CODE REVIEW

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [inline kernel/src/log_segment.rs:457]: I wanted to flag this section. We are effectively doing the same work as the `build` function would do. However, `build` maps the `ParsedLogPath` to `file.location`. We need more information in this test than location, so we can't `build` and make the same assertions. 

I considered factoring out part of `build` into a `process_files` function, but it didn't seem worth adding an entire extra function when this is the only test that needs it.

File: kernel/src/snapshot.rs
Comment (OussamaSaoudi-db) [inline kernel/src/snapshot.rs:31]: I also wanted to flag this part. The version is just the `end_version` or effective version of the log segment. This seemed to me a property of `LogSegment` rather than a property of `Snapshot`. Let me know what you think :)

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [inline kernel/src/log_segment.rs:256]: Is the `max_checkpoint_version = -1` sensible? This is what we used to do in the checkpoint case, but perhaps we want to rethink this?

**Comment by OussamaSaoudi-db** [comment]
@nicklan @scovich @zachschuermann  I was working on CDF and needed a way to keep track of the commit version. Currently, `LogSegment` holds `commit_files: Vec<FileMeta>`, and `checkpoint_files: Vec<FileMeta>`. This only holds the path and modification time.  It would be really useful to keep all the commit version in `ParsedLogFile`.  Here are some alternatives I considered:
- `commit_files.iter().zip(start_version..=end_version)` to keep track of versions when iterating over files. I feel like this could be error prone, but nothing some checks wouldn't fix
- parse the commit version from the `commit_file.path`. This feels icky.


How do you feel about `LogSegment` having `commit_files: Vec<ParsedLogPath>` instead of the current `commit_files: Vec<FileMeta>`? The opportunities I see are the following:
1. CDF can directly get the commit version from the `ParsedLogPath`. We're certain this is the right commit number.
2. The test I flagged above wants to have access to `version`, but we can't use `build` because that projects away version info. I think keeping version info and `file_type` from `ParsedLogPath` improves the testability of `LogSegment` and by extension, `Snapshot`.
3. If we want to have logging and metrics observability, knowing commit version and file type may be useful. Consider a scenario where we want to monitor performance, and we notice an anomaly when replaying the log. It would be handy to instantly know the version number or file type instead of having to parse through the path name.

The main drawbacks I see is that we're stuck moving around more data, and the `LogPathFileType` enum may not be FFI friendly.  This is also a relatively larger change, so maybe we can defer this and do the alternatives I considered above.

**Comment by zachschuermann** [comment]
Hey @OussamaSaoudi-db thanks for that message I broadly agree with your ideas and (without too much digging yet) sounds like doing both `commit_files` and `checkpoint_files` as vecs of `ParsedLogPath` makes a lot of sense. Perhaps we can do that as a quick pre-factor to this PR?

Regarding the downsides:
> The main drawbacks I see is that we're stuck moving around more data, and the `LogPathFileType` enum may not be FFI friendly. This is also a relatively larger change, so maybe we can defer this and do the alternatives I considered above.

I doubt the marginal increase in size of the structs themselves will have meaningful impact (or, put another way, whatever impact is likely worth it). Do we expose `LogSegment`s in FFI? right now it's `pub(crate)` with `developer-visibility` for `pub` right?

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:134]: do we ever want non-ordered commit files?

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:133]: nit
```suggestion
            skip_checkpoint_files: false,
```

Comment (zachschuermann) [reply kernel/src/log_segment.rs:133]: or 
```suggestion
            omit_checkpoint_files: false,
```

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:118]: maybe this should be `table_root` if we are appending `_delta_log` to it? I recall not liking how many of these paths were handled but that's another issue for me to dig up and we can fix later :)

File: kernel/src/log_segment.rs
Comment (zachschuermann) [reply kernel/src/log_segment.rs:457]: and i suppose if we have `LogSegment` hold `ParsedLogPath`s then this is moot?

File: kernel/src/snapshot.rs
Comment (zachschuermann) [reply kernel/src/snapshot.rs:31]: yea if we do this can we have a `snapshot.version()` getter then? to just pass through the logic to ask it's log segment? I think in general we want to maintain the notion of a snapshot 'having a version'

EDIT you already have it :) this sounds good then!

**Review by zachschuermann — COMMENTED** [review]
flushing some comments. generally this looks great, I think the separation really makes sense! you think we can relatively easily do a pre-factor PR to change `LogSegment` to track `ParsedLogPath`s?

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs:134]: So this is a bit of a naming issue. The default behaviour for snapshot is to return commit files in reverse order. I made this the default for `LogSegment`.  Maybe I should instead make it `reverse_commit_files` and make `Snapshot` explicitly ask for reverse.

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs:24]: This is what I'd proposed. I moved `end_version` to `LogSegment`. The flag above was just double checking that this makes sense.

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs]: Correct. Just updated the code to use `ParsedLogPath` and update the test.

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs]: I went ahead and switched it to `with_reversed_commit_files` PTAL.

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs]: Good point. Changed it to be called `table_root`.

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:12]: nit: i think the large braces in imports actually worsen readability - can we revert?

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:17]: same as above
```suggestion
use std::cmp::Ordering;
use std::sync::{Arc, LazyLock};
```

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:142]: does this work?
```suggestion
            ..Default::default()
```

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:152]:
```suggestion
    /// and checkpoints are at this version or above it.
```

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:124]: nit: docs; can be something short/simple like "builder for log segment from [start_version, end_version] _inclusive_!"

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:215]: mega nit but this feels more understandable to me
```suggestion
            commit_files.retain(|log_path| log_path.version >= start_version);
```

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:159]:
```suggestion
    /// and checkpoints are at or below the end version.
```

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:158]:
```suggestion
    /// Optionally set the end version (inclusive) of the [`LogSegment`]. This ensures that all commit files
```

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:151]:
```suggestion
    /// Optionally set the start version (inclusive) of the [`LogSegment`]. This ensures that all commit files
```

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:167]: do we need this? considering we expose the commit files and checkpoint files separately can't the consumer just elect to use one but not the other?

**Review by zachschuermann — COMMENTED** [review]
flushing comments will finish up soon

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs:167]: Yeah good point.  This may have been an inelegant way of solving this problem: Commit files do not intersect with checkpoints.  Consider the following: 
```
if omit_checkpoint_files {
    checkpoint_files.clear();
}
// ...
if let Some(checkpoint_file) = checkpoint_files.first() {
    commit_files.retain(|log_path| checkpoint_file.version < log_path.version);
}
```
We ensure commit versions and checkpoint versions are disjoint  by `retain`ing commit files that are greater than a checkpoint version. This is the expected behaviour in `Snapshot`. 


Now that I've written and thought about this, I think I can go ahead and do this:
```
    commit_files.retain(|log_path| checkpoint_file.version < log_path.version || log_path.version > start_version);

```

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs:142]: Darn unfortunately not 😔

File: kernel/src/log_segment.rs
Comment (zachschuermann) [reply kernel/src/log_segment.rs:262]: why not just `Option<Version>`?

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:321]:
```suggestion
        if max_checkpoint_version != checkpoint_metadata.version.try_into()? {
```

Comment (zachschuermann) [reply kernel/src/log_segment.rs:321]: can we just make max_checkpoint_version a `Version`? or option version? which is u64

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:355]: nit: unnest

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:221]: this relies on checkpoint ordering?

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:197]: maybe rename to `sorted_commit_files`/`sorted_checkpoint_files`

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:199]: I think this should be 
```suggestion
            (Some(cp), Some(end_version)) if cp.version <= end_version => {
```

Comment (zachschuermann) [reply kernel/src/log_segment.rs:199]: (i think this was existing bug)

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:400]: can we add a few tests:
1. fail for logsegment that's not continuous
2. check on a CDF-style log segment (build one with omit_checkpoints = true) that has checkpoints in the range
3. sorting tests for commits and checkpoints
4. check for LogSegment created with checkpoint hint (to check that bug above where I switched `>=` to `<= cp.version`

**Review by zachschuermann — COMMENTED** [review]
left a bunch of comments. i think we are getting close but I'd like to reconsider having some sort of bifurcation for a snapshot segment vs. a cdf segment.

when we do listing, I don't think it will work for a CDF segment. if we have a `start_version` and a `start_checkpoint` set, then won't we do a list from `start_checkpoint` which will miss versions between `start_version..start_checkpoint`? perhaps we can just work around this for now by disallowing both to be set?

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:275]: future work: peek then push to validate ordering/gaps?

File: kernel/src/log_segment.rs
Comment (zachschuermann) [reply kernel/src/log_segment.rs:264]: should we rename to `checkpoint_version`? and probably have docs that says that we are really only tracking one checkpoint (the highest) and N commits

File: kernel/src/log_segment.rs
Comment (zachschuermann) [reply kernel/src/log_segment.rs:40]: i know this is a breaking change but I think we may want to consider calling this `checkpoint_parts`. check my understanding here, but I think this is only ever really supposed to capture a single which may be multiple parts

Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs:40]: You're right, the multiple files are just multiple parts. It's mentioned in `list_log_files_with_checkpoint`:
```
        } else if checkpoint_files.len() != checkpoint_metadata.parts.unwrap_or(1) {
            return Err(Error::Generic(format!(
                "_last_checkpoint indicated that checkpoint should have {} parts, but it has {}",
                checkpoint_metadata.parts.unwrap_or(1),
                checkpoint_files.len()
            )));
        }
```

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs:275]: I like that 👌

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs:223]: This doesn't. Thinking about it some more, I'm not sure there is a notion of checkpoint ordering? They all have the same version. Perhaps it's ordered by part number, but I haven't seen anything to indicate that.

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs]: For context, we decided to go with `omit_checkpoint_files` in the end. It was a more explicit approach.

File: kernel/src/log_segment.rs
Comment (nicklan) [inline kernel/src/log_segment.rs:139]: this is for the builder right, so it should be an instruction, suggest:
```suggestion
    sort_commit_files_ascending: bool,
```

File: kernel/src/log_segment.rs
Comment (nicklan) [inline kernel/src/log_segment.rs:265]: I'm unclear why this method and the one below are in the `impl` for `LogSegmentBuilder`? I guess you want to call them as `Self::xxx` above, but I don't think that gives any benefit, and I'd just move them out to stand-alone methods

File: kernel/src/log_segment.rs
Comment (nicklan) [inline kernel/src/log_segment.rs:143]: I'd say, let's get rid of the `fs_client` and `table_root` args, and not have them be part of the struct. Then `build()` can just take them as args. Then we don't need to mess with lifetimes here at all.

File: kernel/src/log_segment.rs
Comment (nicklan) [reply kernel/src/log_segment.rs:412]: Yeah, a few more tests would be good. Also:
- Errors if start_version and start_checkpoint are both specified

**Review by nicklan — CHANGES_REQUESTED** [review]

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs:175]: Since we error out in this case, this is no longer relevant. Setting to resolved.

File: kernel/src/log_segment.rs
Comment (nicklan) [inline kernel/src/log_segment.rs:446]: you can use the  brand new `delta_path_for_version` in test-utils. You can delete this method and replace all calls to it with calls to that.

**Review by nicklan — APPROVED** [review]
lgtm with one change you should make

**Review by zachschuermann — COMMENTED** [review]
few final comments - can we also add a test with a `start_checkpoint` and an `end_version`

**Comment by zachschuermann** [comment]
Oh and note that `LogSegmentBuilder` is `pub(crate)` with developer-visibility `pub` but everything in it is just `pub(crate)` maybe we just make it all only `pub(crate)` for now?

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs:278]: Version contiguity is important to the properties of a `LogSegment`.  The debug_assert only checks that fs client returns sorted commit files. Resolving for now :)

File: kernel/src/log_segment.rs
Comment (zachschuermann) [inline kernel/src/log_segment.rs:40]: nit:
```suggestion
    /// Sorted commit files in the log segment (see builder for sorting)
    pub commit_files: Vec<ParsedLogPath>,
```

and probably in the future we want to propagate sort information down to there but let's take that as follow-up

**Review by zachschuermann — APPROVED** [review]
LGTM thanks for taking some time for this! did we add test from above for a `start_checkpoint` and an `end_version` check?

File: kernel/src/log_segment.rs
Comment (scovich) [inline kernel/src/log_segment.rs:316]: This method was just moved (unchanged) from snapshot.rs?
Same for `list_log_files_with_checkpoint` below?

File: kernel/src/log_segment.rs
Comment (zachschuermann) [reply kernel/src/log_segment.rs:156]: Hey @scovich this is great thanks for the thorough ideas here - do you think I can take this as a follow-up after this gets merged just to unblock @OussamaSaoudi-db's CDF work? I was thinking of something of this sort but ultimately didn't want to derail a ton an spend more time on this. Hopefully these changes will just require only minor modifications to how `LogSegment`s are created so seems like I could work on this as @OussamaSaoudi-db moves on.

Regarding the actual implementation for this: I think I totally agree on everything here - but additionally, I'm wondering if we want two different _kinds_ of `LogSegment`s. We know that a `LogSegment` is some continuous chunk of commits/checkpoint but we have different invariants based on the kind right? CDF `LogSegment`s _only_ ever have commits, whereas snapshot `LogSegment`s have either commits/checkpoint or only commits. and ordering could/should be different between them?

We could still have a unified `LogSegment` with `sorted_commit_files`, `checkpoint_parts`, and `commit_file_sort_order` which I think would capture both, but want to bring up the possibility of an `enum LogSegment` which has `LogSegment::CDF` and `LogSegment::Snapshot`? but maybe that's getting overly specific/specialized and we want to keep it more general? curious to hear your thoughts :)

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs:167]: I like the `impl Into<Option<Foo>>` approach!

File: kernel/src/log_segment.rs
Comment (OussamaSaoudi-db) [reply kernel/src/log_segment.rs:316]: `list_log_files_with_checkpoint` and `list_log_files_with_version` actually had much of the same code twice. I've merged the two so that you list and process files in one place.

File: kernel/src/log_segment.rs
Comment (zachschuermann) [reply kernel/src/log_segment.rs:156]: issue here: https://github.com/delta-incubator/delta-kernel-rs/issues/476

Comment (zachschuermann) [reply kernel/src/log_segment.rs:156]: Oh and yea I agree reversing the vec on consumer side is probably better - just read that part

File: kernel/src/log_segment.rs
Comment (scovich) [reply kernel/src/log_segment.rs:72]: Ohhhh! I worried for a moment that we were passing just file names. You're right tho, `location` is a `FileMeta` now in this PR.

**Comment by zachschuermann** [comment]
done in #495

**Pull request closed without merging**
